std_reference_basic/src/contract.rs

```diff
@@ -58,207 +28,189 @@ pub fn execute(
     }
 
-    Ok(Response::default())
+    Ok(Response::new().add_attribute("action", "add_relayers"))
 }
 
-pub fn try_remove_relayers(
+pub fn execute_remove_relayers(
     deps: DepsMut,
     info: MessageInfo,
     relayers: Vec<Addr>,
 ) -> StdResult<Response> {
-    let api = deps.api;
-    let owner_addr = read_owner_store(deps.storage).load()?;
-    if api.addr_canonicalize(&info.sender.as_str())? != owner_addr {
+    let config = CONFIG.load(deps.storage)?;
+    if info.sender != config.owner {
         return Err(StdError::generic_err("NOT_AUTHORIZED"));
     }
 
-    let mut relayers_store = relayers_store(deps.storage);
-    for relayer in relayers {
-        relayers_store.set(
-            relayer.to_string().as_bytes(),
-            &bincode::serialize(&false).unwrap(),
-        );
+    for relayer_addr in relayers {
+        RELAYERS.remove(deps.storage, &relayer_addr.to_string());
     }
 
-    Ok(Response::default())
+    Ok(Response::new().add_attribute("action", "remove_relayers"))
 }
 
-pub fn try_relay(
+pub fn execute_relay(
     deps: DepsMut,
     info: MessageInfo,
     symbols: Vec<String>,
     rates: Vec<Uint128>,
     resolve_time: u64,
     request_id: u64,
 ) -> StdResult<Response> {
-    let size = symbols.len();
-    if !(rates.len() == size) {
-        return Err(StdError::generic_err("NOT_ALL_INPUT_SIZES_ARE_THE_SAME"));
-    }
-
-    let is_relayer = query_is_relayer(deps.as_ref(), info.sender.clone())?;
-    if !is_relayer {
+    if !query_is_relayer(deps.as_ref(), info.sender).unwrap() {
         return Err(StdError::generic_err("NOT_A_RELAYER"));
     }
 
-    let mut refs = ref_data_store(deps.storage);
-    for (s, r) in zip!(&symbols, &rates) {
-        let ref_data = refs
-            .get(&s.as_bytes())
-            .map(|b| bincode::deserialize::<RefData>(&b).unwrap());
+    if !(rates.len() == symbols.len()) {
+        return Err(StdError::generic_err("MISMATCHED_INPUT_SIZES"));
+    }
 
-        if ref_data.is_none() || resolve_time > ref_data.unwrap().resolve_time {
-            refs.set(
-                s.as_bytes(),
-                &bincode::serialize(&RefData::new(r, resolve_time, request_id)).unwrap(),
-            );
+    for (symbol, rate) in symbols.into_iter().zip(rates.into_iter()) {
+        match REFDATA.may_load(deps.storage, &symbol)? {
+            Some(existing_refdata) => {
+                if existing_refdata.resolve_time < resolve_time {
+                    REFDATA.save(
+                        deps.storage,
+                        &symbol,
+                        &RefData::new(rate, resolve_time, request_id),
+                    )?;
+                } else {
+                    return Err(StdError::generic_err("INVALID_RESOLVE_TIME"));
+                }
+            }
+            None => REFDATA.save(
+                deps.storage,
+                &symbol,
+                &RefData::new(rate, resolve_time, request_id),
+            )?,
```

Comment:
Let's refactor to pattern if there is exist data. check resolve time and if old just skip this symbol, then save data to store.